### PR TITLE
Add TLS 1.3 Support to WinSim Demos

### PR DESCRIPTION
**Pull request by Skptak**
Description
-----------
Update the submodule for MbedTLS to their [v3.5.1](https://github.com/Mbed-TLS/mbedtls/tree/v3.5.1) release
Change the `mbedtls_config_v3.2.1.h` file to be `mbedtls_config_v3.5.1.h`
Add very basic support for [MbedTLS PSA](https://mbed-tls.readthedocs.io/en/latest/getting_started/psa/) by adding calls to 'psa_crypto_init()' to the mbedtls transport layers
Add support for TLS 1.3 by using `MBEDTLS_SSL_PROTO_TLS1_3`
Slight tweaks to some demos to account for the changes.
Add `WIN32_LEAN_AND_MEAN` to WinSim Demo configurations as required by MBedTLS's [x509_crt.c and other files](https://github.com/Mbed-TLS/mbedtls/pull/7436#discussion_r1171279453)

Modify the Fleet Provisioning Demo scripts to use ECDSA keys, documentation update for website has also been made.

Test Steps
-----------
All demos were run locally using Visual Studio 2019
All demos were run using internal CI-CD Infrastructure

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_S3_Download/demo_config.h [hunk 1]
```diff
@@ -207,7 +207,7 @@ extern void vLoggingPrintf( const char * pcFormatString,
 /**
  * @brief Transport timeout in milliseconds for transport send and receive.
  */
-#define democonfigTRANSPORT_SEND_RECV_TIMEOUT_MS    ( 5000 )
+#define democonfigTRANSPORT_SEND_RECV_TIMEOUT_MS    ( 10000 )
```
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_S3_Download/demo_config.h:210]: The 5 second timeout was causing the demo to fail when using debug printing, bump it to 10 second timeout to create a more forgiving experience.

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Basic_TLS/DemoTasks/BasicTLSMQTTExample.c [hunk 2]
```diff
@@ -125,7 +125,7 @@
 /**
  * @brief Timeout for receiving CONNACK packet in milliseconds.
  */
-#define mqttexampleCONNACK_RECV_TIMEOUT_MS                ( 1000U )
+#define mqttexampleCONNACK_RECV_TIMEOUT_MS                ( 2000U )
```
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Basic_TLS/DemoTasks/BasicTLSMQTTExample.c:128]: Timeout increase to account for debug prints causing demo failure.

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Basic_TLS/DemoTasks/BasicTLSMQTTExample.c [hunk 3]
```diff
@@ -160,7 +160,7 @@
  * @brief Timeout for MQTT_ProcessLoop in milliseconds.
  * Refer to FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/readme.txt for more details.
  */
-#define mqttexamplePROCESS_LOOP_TIMEOUT_MS                ( 2000U )
+#define mqttexamplePROCESS_LOOP_TIMEOUT_MS                ( 5000U )
```
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Basic_TLS/DemoTasks/BasicTLSMQTTExample.c:163]: Timeout increase to account for debug prints causing demo failure.

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Basic_TLS/DemoTasks/BasicTLSMQTTExample.c [hunk 4]
```diff
@@ -443,11 +443,12 @@ void vStartSimpleMQTTDemo( void )
      * state or call the MQTT_ProcessLoop() API function. Using an agent task
      * also enables multiple application tasks to more easily share a single
      * MQTT connection. */
+
     xTaskCreate( prvMQTTDemoTask,          /* Function that implements the task. */
                  "DemoTask",               /* Text name for the task - only used for debugging. */
                  democonfigDEMO_STACKSIZE, /* Size of stack (in words, not bytes) to allocate for the task. */
                  NULL,                     /* Task parameter - not used in this case. */
-                 tskIDLE_PRIORITY,         /* Task priority, must be between 0 and configMAX_PRIORITIES - 1. */
+                 tskIDLE_PRIORITY + 1,     /* Task priority, must be between 0 and configMAX_PRIORITIES - 1. */
```
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Basic_TLS/DemoTasks/BasicTLSMQTTExample.c:451]: Increase prio of the demo task to ensure it does not get starved by the logging task.

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Mutual_Auth/MQTT_Mutual_Auth.vcxproj [hunk 5]
```diff
@@ -52,9 +52,9 @@
...
+      <PreprocessorDefinitions>MBEDTLS_CONFIG_FILE="mbedtls_config_v3.5.1.h";_CRT_SECURE_NO_WARNINGS;WIN32;_DEBUG;_CONSOLE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <ConformanceMode>true</ConformanceMode>
-      <AdditionalIncludeDirectories>.\;..\Common;..\..\..\Source\Application-Protocols\network_transport\tcp_sockets_wrapper\include;..\..\..\Source\Application-Protocols\network_transport;..\..\..\Source\Utilities\backoff_algorithm\source\include;..\..\..\Source\Application-Protocols\coreMQTT\source\include;..\..\..\Source\Application-Protocols\coreMQTT\source\interface;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>.\;..\..\..\VisualStudio_StaticProjects\MbedTLS;..\Common;..\..\..\Source\Application-Protocols\network_transport\tcp_sockets_wrapper\include;..\..\..\Source\Application-Protocols\network_transport;..\..\..\Source\Utilities\backoff_algorithm\source\include;..\..\..\Source\Application-Protocols\coreMQTT\source\include;..\..\..\Source\Application-Protocols\coreMQTT\source\interface;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Mutual_Auth/MQTT_Mutual_Auth.vcxproj:57]: Need to include this directory so the project can include the mbedtls_config file

File: FreeRTOS-Plus/Source/Application-Protocols/network_transport/mbedtls_pk_pkcs11.c [hunk 6]
```diff
@@ -92,7 +110,7 @@ static void * p11_ecdsa_ctx_alloc( void );
  * @param xPkHandle The CK_OBJECT_HANDLE for the target private key.
  * @return CKR_OK on success
  */
-static CK_RV p11_ecdsa_ctx_init( void * pvCtx,
+static CK_RV p11_ecdsa_ctx_init( mbedtls_pk_context * pxMbedtlsPkCtx,
```
Comment (Skptak) [inline FreeRTOS-Plus/Source/Application-Protocols/network_transport/mbedtls_pk_pkcs11.c:113]: There was a change to these functions to take in the `mbedtls_pk_context` instead of taking in `mbedtls_pk_context->pk_ctx` link to the [PR](https://github.com/Mbed-TLS/mbedtls/pull/7463)

File: .github/scripts/core_checker.py [hunk 7]
```diff
@@ -258,7 +258,7 @@
 
 FREERTOS_IGNORED_PATTERNS = [
     r'.*\.git.*',
-    r'.*mbedtls_config\.h.*',
+    r'.*mbedtls_config*',
```
Comment (Skptak) [inline .github/scripts/core_checker.py:261]: Change how we exclude the mbedtls_config file from this check to ignore the pattern of the file VS calling it out explicitly.

File: .gitmodules [hunk 8]
```diff
@@ -24,7 +24,8 @@
 	url = https://github.com/Mbed-TLS/mbedtls.git
 [submodule "FreeRTOS-Plus/Source/Application-Protocols/corePKCS11"]
 	path = FreeRTOS-Plus/Source/corePKCS11
-	url = https://github.com/FreeRTOS/corePKCS11.git
+	url = https://github.com/skptak/corePKCS11.git
+	branch = mbedtls3_5_0Upgrade
```
Comment (Skptak) [inline .gitmodules:28]: These changes still need to be merged.

File: FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/DemoSetup/convert_credentials_to_der.py [hunk 9]
```diff
@@ -6,8 +6,8 @@
 from cryptography.hazmat.backends import default_backend
 from cryptography.hazmat.primitives import serialization
 
-KEY_OUT_NAME = f"{os.getcwd()}\\corePKCS11_Claim_Key.dat"
-CERT_OUT_NAME = f"{os.getcwd()}\\corePKCS11_Claim_Certificate.dat"
+KEY_OUT_NAME = os.path.join(os.getcwd(), 'corePKCS11_Claim_Key.dat')
+CERT_OUT_NAME = os.path.join(os.getcwd(), 'corePKCS11_Claim_Certificate.dat')
```
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/DemoSetup/convert_credentials_to_der.py:10]: Currently these scripts only work on windows command line (or powershell), slight tweak to make them work with unix terminals as well for developers who use WSL.

File: FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/DemoSetup/demo_setup.py [hunk 10]
```diff
@@ -6,8 +6,8 @@
 import botocore
 from convert_credentials_to_der import convert_pem_to_der
 
-KEY_OUT_NAME = f"{os.getcwd()}\\corePKCS11_Claim_Key.dat"
-CERT_OUT_NAME = f"{os.getcwd()}\\corePKCS11_Claim_Certificate.dat"
+KEY_OUT_NAME = os.path.join(os.getcwd(), "corePKCS11_Claim_Key.dat")
+CERT_OUT_NAME = os.path.join(os.getcwd(), "corePKCS11_Claim_Certificate.dat")
```
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/DemoSetup/demo_setup.py:10]: Same as [above comment](https://github.com/FreeRTOS/FreeRTOS/pull/1135/files#r1422817806)

File: FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/FleetProvisioningDemoExample.c [hunk 11]
```diff
@@ -646,6 +646,12 @@ int prvFleetProvisioningTask( void * pvParameters )
              * topics. In this demo we use CBOR encoding for the payloads,
              * so we use the CBOR variants of the topics. */
             xStatus = prvSubscribeToCsrResponseTopics();
+
+            if( xStatus == true )
+            {
+                /* Subscribe to the RegisterThing response topics. */
+                xStatus = prvSubscribeToRegisterThingResponseTopics();
+            }
```
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/FleetProvisioningDemoExample.c:655]: Per documentation on AWS for the [Device Provisioning MQTT API](https://docs.aws.amazon.com/iot/latest/developerguide/fleet-provision-api.html) subscribe to relevant topics before publishing to them.


> Important
> Before publishing a request message topic, subscribe to the response topics to receive the response. The messages used by this API use MQTT's publish/subscribe protocol to provide a request and response interaction.
> If you do not subscribe to the response topics before you publish a request, you might not receive the results of that request.

File: FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/FleetProvisioningDemoExample.c [hunk 12]
```diff
@@ -124,29 +124,30 @@
...
+ * @note Has a maximum length of 64 for more information see the following link
+ * https://docs.aws.amazon.com/iot/latest/apireference/API_Certificate.html#iot-Type-Certificate-certificateId
  */
 #define fpdemoCERT_ID_BUFFER_LENGTH                          64
 
 /**
  * @brief Size of buffer in which to hold the certificate ownership token.
  */
-#define fpdemoOWNERSHIP_TOKEN_BUFFER_LENGTH                  512
+#define fpdemoOWNERSHIP_TOKEN_BUFFER_LENGTH                  1024
```
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/FleetProvisioningDemoExample.c:150]: Use of an ECDSA key can require a larger buffer to hold some of this info, double the size of these buffers to account for this change

File: FreeRTOS-Plus/Demo/AWS/Device_Defender_Windows_Simulator/Device_Defender_Demo/Device_Defender_Demo.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Device_Defender_Windows_Simulator/Device_Defender_Demo/Device_Defender_Demo.vcxproj:1]: Follow up of #1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/AWS/Device_Shadow_Windows_Simulator/Device_Shadow_Demo/Device_Shadow_Demo.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Device_Shadow_Windows_Simulator/Device_Shadow_Demo/Device_Shadow_Demo.vcxproj:1]: Follow up of #1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/AWS/Device_Shadow_Windows_Simulator/Device_Shadow_Demo/DemoTasks/ShadowDemoMainExample.c [hunk 13]
```diff
@@ -390,7 +390,7 @@ static BaseType_t prvWaitForDeleteResponse( MQTTContext_t * pxMQTTContext )
         /* Event callback will set #xDeleteResponseReceived when receiving an
          * incoming publish on either `/delete/accepted` or `/delete/rejected`
          * Shadow topics. */
-        xMQTTStatus = MQTT_ProcessLoop( pxMQTTContext, MQTT_PROCESS_LOOP_TIMEOUT_MS );
+        xMQTTStatus = MQTT_ProcessLoop( pxMQTTContext );
```
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Device_Shadow_Windows_Simulator/Device_Shadow_Demo/DemoTasks/ShadowDemoMainExample.c:393]: Signature for this function has changed, but the incorrect parameter here isn't causing build failures.

Link to mainline coreMQTT definition of this function is [here](https://github.com/FreeRTOS/coreMQTT/blob/main/source/core_mqtt.c#L3121)
```
MQTTStatus_t MQTT_ProcessLoop( MQTTContext_t * pContext )
```

File: FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/fleet_provisioning_demo.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/fleet_provisioning_demo.vcxproj:1]: Follow up of #1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/AWS/Jobs_Windows_Simulator/Jobs_Demo/Jobs_Demo.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Jobs_Windows_Simulator/Jobs_Demo/Jobs_Demo.vcxproj:1]: Follow up of #1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/AWS/Ota_Windows_Simulator/Ota_Over_Http_Demo/Ota_Over_Http_Demo.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Ota_Windows_Simulator/Ota_Over_Http_Demo/Ota_Over_Http_Demo.vcxproj:1]: Follow up of #1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/AWS/Ota_Windows_Simulator/Ota_Over_Mqtt_Demo/Ota_Over_Mqtt_Demo.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Ota_Windows_Simulator/Ota_Over_Mqtt_Demo/Ota_Over_Mqtt_Demo.vcxproj:1]: Follow up of #1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/FreeRTOS_Cellular_Interface_Windows_Simulator/MQTT_Mutual_Auth_Demo_with_SARA_R4/mqtt_mutual_auth_demo_with_sara_r4.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/FreeRTOS_Cellular_Interface_Windows_Simulator/MQTT_Mutual_Auth_Demo_with_SARA_R4/mqtt_mutual_auth_demo_with_sara_r4.vcxproj:1]: Follow up of https://github.com/FreeRTOS/FreeRTOS/pull/1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_CLI_with_Trace_Windows_Simulator/WIN32.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_CLI_with_Trace_Windows_Simulator/WIN32.vcxproj:1]: Dangling file that was not used, removing it.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_CLI_with_Trace_Windows_Simulator/WIN32.vcxproj.filters
Comment (Skptak) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_CLI_with_Trace_Windows_Simulator/WIN32.vcxproj.filters:1]: Dangling file that was not used. Removing it.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_WolfSSL_Windows_Simulator/FreeRTOS_Plus_WolfSSL_Windows_Simulator.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_WolfSSL_Windows_Simulator/FreeRTOS_Plus_WolfSSL_Windows_Simulator.vcxproj:1]: Follow up of https://github.com/FreeRTOS/FreeRTOS/pull/1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_Mutual_Auth/CoreHTTP_Mutual_Auth.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_Mutual_Auth/CoreHTTP_Mutual_Auth.vcxproj:1]: Follow up of https://github.com/FreeRTOS/FreeRTOS/pull/1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_Plaintext/CoreHTTP_Plaintext.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_Plaintext/CoreHTTP_Plaintext.vcxproj:1]: Follow up of https://github.com/FreeRTOS/FreeRTOS/pull/1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_S3_Download/CoreHTTP_S3_Download.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_S3_Download/CoreHTTP_S3_Download.vcxproj:1]: Follow up of https://github.com/FreeRTOS/FreeRTOS/pull/1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_S3_Download_Multithreaded/CoreHTTP_S3_Download_Multithreaded.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_S3_Download_Multithreaded/CoreHTTP_S3_Download_Multithreaded.vcxproj:1]: Follow up of https://github.com/FreeRTOS/FreeRTOS/pull/1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_S3_Upload/CoreHTTP_S3_Upload.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_S3_Upload/CoreHTTP_S3_Upload.vcxproj:1]: Follow up of https://github.com/FreeRTOS/FreeRTOS/pull/1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Serializer/MQTT_Serializer.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Serializer/MQTT_Serializer.vcxproj:1]: Follow up of https://github.com/FreeRTOS/FreeRTOS/pull/1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/fleet_provisioning_demo.vcxproj
Comment (Skptak) [inline FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/fleet_provisioning_demo.vcxproj:1]: Follow up of #1120, some projects have dangling selections for non-existing configurations.

File: FreeRTOS-Plus/VisualStudio_StaticProjects/FreeRTOS-Kernel/FreeRTOSConfig.h [hunk 14]
```diff
@@ -139,7 +139,7 @@ void vConfigureTimerForRunTimeStats( void );
  * results in the wired network being used, while setting
  * configNETWORK_INTERFACE_TO_USE to 2 results in the wireless network being
  * used. */
-#define configNETWORK_INTERFACE_TO_USE      ( 1L )
+#define configNETWORK_INTERFACE_TO_USE      ( 4L )
```
Comment (bradleysmith23) [inline FreeRTOS-Plus/VisualStudio_StaticProjects/FreeRTOS-Kernel/FreeRTOSConfig.h:142]: nit: Should this stay as 1?

**Review by bradleysmith23 — DISMISSED** [review]
Approved, left one small comment

File: FreeRTOS-Plus/VisualStudio_StaticProjects/FreeRTOS-Kernel/FreeRTOSConfig.h (hunk 14 above)
Comment (Skptak) [reply FreeRTOS-Plus/VisualStudio_StaticProjects/FreeRTOS-Kernel/FreeRTOSConfig.h:142]: I think we should move it to 4?
As stated it ends up being the actual Ethernet port on some laptops. Then it was 4 for my virtual windows environment.

Where users will likely need to come in and change this anyways, but maybe this value happens to be correct more often?

**Review by bradleysmith23 — DISMISSED** [review]

File: FreeRTOS-Plus/VisualStudio_StaticProjects/FreeRTOS-Kernel/FreeRTOSConfig.h (hunk 14 above)
Comment (AniruddhaKanhere) [reply FreeRTOS-Plus/VisualStudio_StaticProjects/FreeRTOS-Kernel/FreeRTOSConfig.h:142]: In this case, I think we should set it to 0. That way the system will print an error with the reason instead of leaving folks to wonder why their demo doesn't work.
See here: https://github.com/FreeRTOS/FreeRTOS/blob/b9cb18e46e7042d1dc1475bf054cc5278e422d16/FreeRTOS-Plus/VisualStudio_StaticProjects/FreeRTOS%2BTCP/NetworkInterface_WinPCap.c#L487

File: FreeRTOS-Plus/VisualStudio_StaticProjects/MbedTLS/mbedtls_freertos_port.c [hunk 15]
```diff
@@ -45,6 +39,13 @@
 
 #include "mbedtls_freertos_port.h"
 
+
+#include "FreeRTOSConfig.h"
+
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "semphr.h"
+
```
Comment (AniruddhaKanhere) [inline FreeRTOS-Plus/VisualStudio_StaticProjects/MbedTLS/mbedtls_freertos_port.c:48]: Curious why this change was required?

File: FreeRTOS-Plus/Source/Application-Protocols/network_transport/mbedtls_pk_pkcs11.c [hunk 16]
```diff
@@ -442,6 +460,7 @@ static CK_RV p11_ecdsa_ctx_init( void * pvCtx,
     else
     {
         xResult = CKR_FUNCTION_FAILED;
+        return xResult;
```
[the remark this answers is not in the document]
Comment (Skptak) [reply FreeRTOS-Plus/Source/Application-Protocols/network_transport/mbedtls_pk_pkcs11.c:463]: I went with the early return because I thought it was cleaner than needing to wrap the rest of the function in the else block?

Where to be fair, this is a MISRA violation but I don't believe we enforce MISRA compliance on these files. But I can do that if you want.

There are many other places that the rules from MISRA are ignored here i.e. almost all of the casting that is done between [different types](https://github.com/FreeRTOS/FreeRTOS/blob/34b2048819a2fd85830e1489a36786ddc603f085/FreeRTOS-Plus/Source/Application-Protocols/network_transport/mbedtls_pk_pkcs11.c#L449):
```
P11EcDsaCtx_t * pxP11EcDsaCtx = ( P11EcDsaCtx_t * ) pk;
```

And the fact that a majority of the function parameters aren't marked `const` when they aren't modified. Where that's because MbedTLS expects these to be marked as non-constant, where making them constant would raise the dropping the `const` tag [compilation warning](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4287?view=msvc-170)

[the remark this answers is not in the document]
Comment (Skptak) [reply FreeRTOS-Plus/Source/Application-Protocols/network_transport/mbedtls_pk_pkcs11.c:463]: > Just musing: With this change, do we still need bigger timeouts?

I know that this could be reverted - I had mainly done this change for two reasons
1. I was originally seeing an issue with the task being starved/delayed when I was using the logging task to print out mbedtls debug info. Where doing this fixed that.
1. But then I thought it felt odd that we were showing examples of every task running at the same priority like this? This just seems more correct in my mind that this task *IS* a higher priority than the logging or idle task.
    * Where I had actually considered adding task priority changes with the other demos.
    * But the scope of this work has already gone well over what was intended, so I didn't do this.

Where all said, I'm fine with reverting this. But I'd prefer to leave it in, since it is inconsistent now.

File: FreeRTOS-Plus/VisualStudio_StaticProjects/MbedTLS/mbedtls_freertos_port.c (hunk 15 above)
Comment (Skptak) [reply FreeRTOS-Plus/VisualStudio_StaticProjects/MbedTLS/mbedtls_freertos_port.c:48]: This partly solved/worked around the `winsock2.h` issue that was created with the new version of MbedTLS

I believe at this point it could be reverted if you want

File: FreeRTOS-Plus/VisualStudio_StaticProjects/FreeRTOS-Kernel/FreeRTOSConfig.h (hunk 14 above)
Comment (Skptak) [reply FreeRTOS-Plus/VisualStudio_StaticProjects/FreeRTOS-Kernel/FreeRTOSConfig.h:142]: > In this case, I think we should set it to 0.
I really like that idea actually, thanks!

File: FreeRTOS-Plus/VisualStudio_StaticProjects/MbedTLS/mbedtls_freertos_port.c (hunk 15 above)
Comment (Skptak) [reply FreeRTOS-Plus/VisualStudio_StaticProjects/MbedTLS/mbedtls_freertos_port.c:48]: I think there was another reason I had wanted to include the MbedTLS config first? But I can't remember if there was an actual reason to or not at this point

Comment (AniruddhaKanhere) [reply FreeRTOS-Plus/VisualStudio_StaticProjects/MbedTLS/mbedtls_freertos_port.c:48]: I think we can keep this change. I was just curious why would we need to do this and to understand whether FreeRTOS headers were including some files which did not work well with the mbedtls files.

File: FreeRTOS-Plus/Source/Application-Protocols/network_transport/mbedtls_pk_pkcs11.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (AniruddhaKanhere) [reply FreeRTOS-Plus/Source/Application-Protocols/network_transport/mbedtls_pk_pkcs11.c:463]: I don't mind the priority change and the timeout change. Lets keep it.

For the early return, I agree that these files are not checked for MISRA compliance but these files should be example of best practices. I also agree that adding else would make the code a bit different, but IMO, it is easier to debug if the function doesn't have a ton of return locations. As far as casting is considered, we cannot do much in those cases except add MISRA justifications saying why are we casting and why is it okay to do so.

Unless you are completely against it and willing to share the reasons, I propose that we stick with single point return.

[the remark this answers is not in the document]
Comment (Skptak) [reply FreeRTOS-Plus/Source/Application-Protocols/network_transport/mbedtls_pk_pkcs11.c]: > Unless you are completely against it and willing to share the reasons, I propose that we stick with single point return.

Only reasoning I have is that I think since we ignore a lot of the MISRA best practices in the file, it feels odd to pick and choose which ones to follow? Where we are also ignoring the whole "if statements should have a corresponding else statement" in this function already.

But I agree with your logic that "When possible stick to best practices" so I went ahead and swapped to the single exit point for this function again

File: FreeRTOS-Plus/VisualStudio_StaticProjects/MbedTLS/mbedtls_freertos_port.c (hunk 15 above)
Comment (Skptak) [reply FreeRTOS-Plus/VisualStudio_StaticProjects/MbedTLS/mbedtls_freertos_port.c]: I actually went ahead and reverted it - I don't think this change is actually needed, so would rather keep the file as is.

**Review by bradleysmith23 — APPROVED** [review]

**Review by AniruddhaKanhere — APPROVED** [review]
